sklearn/ensemble/_gb.py

```diff
@@ -23,33 +23,246 @@
 
-from ..utils import check_array, check_random_state, column_or_1d
+from .._loss.loss import (
+    _LOSSES,
+    AbsoluteError,
+    BaseLoss,
+    ExponentialLoss,
+    HalfBinomialLoss,
+    HalfMultinomialLoss,
+    HalfSquaredError,
+    HuberLoss,
+    PinballLoss,
+)
+from ..base import ClassifierMixin
+from ..base import RegressorMixin
+from ..base import is_classifier
+from ..dummy import DummyRegressor, DummyClassifier
+from ..exceptions import NotFittedError
+from ..preprocessing import LabelEncoder
+from ..tree._tree import TREE_LEAF
+from ..utils import check_random_state
+from ..utils import check_array
+from ..utils import column_or_1d
 from ..utils._param_validation import HasMethods, Interval, StrOptions
+from ..utils.fixes import percentile
 from ..utils.validation import check_is_fitted, _check_sample_weight
 from ..utils.multiclass import check_classification_targets
-from ..exceptions import NotFittedError
+from ..utils.stats import _weighted_percentile
+
+from ._base import BaseEnsemble
+from ._gradient_boosting import predict_stages
+from ._gradient_boosting import predict_stage
+from ._gradient_boosting import _random_sample_mask
+
+
+_LOSSES = _LOSSES.copy()
+_LOSSES.update(
+    {
+        "quantile": PinballLoss,
+        "huber": HuberLoss,
+    }
+)
+
+
+def _init_raw_predictions(X, estimator, loss, use_predict_proba):
+    """Return the initial raw predictions.
+
+    Parameters
+    ----------
+    X : ndarray of shape (n_samples, n_features)
+        The data array.
+    estimator : object
+        The estimator to use to compute the predictions.
+    loss : BaseLoss
+        An instace of a loss function class.
+    use_predict_proba : bool
+        Whether estimator.predict_proba is used instead of estimator.predict.
+
+    Returns
+    -------
+    raw_predictions : ndarray of shape (n_samples, K)
+        The initial raw predictions. K is equal to 1 for binary
+        classification and regression, and equal to the number of classes
+        for multiclass classification. ``raw_predictions`` is casted
+        into float64.
+    """
+    # TODO: Use loss.fit_intercept_only where appropriate instead of
+    # DummyRegressor.
+    if use_predict_proba:
+        try:
+            predictions = estimator.predict_proba(X)
+        except AttributeError as e:
+            msg = "The given init estimator does not provide a predict_proba method."
+            raise AttributeError(msg) from e
+        if not loss.is_multiclass:
+            predictions = predictions[:, 1]  # probability of positive class
+        eps = np.finfo(np.float32).eps  # FIXME: This is quite large!
+        predictions = np.clip(predictions, eps, 1 - eps, dtype=np.float64)
+    else:
+        predictions = estimator.predict(X).astype(np.float64)
+
+    if predictions.ndim == 1:
+        return loss.link.link(predictions).reshape(-1, 1)
+    else:
+        return loss.link.link(predictions)
+
+
+def _update_terminal_regions(
```

Comment:
The original `_gb_losses.py` had this abstraction in the loss classes themselves. Given that this is very specific to trees, I am okay with having a big function handle it.

Reply:
I'm also not a big fan of this, but it got the things done. Maybe, someone clever can refactor later in a more beautiful way.